core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java

```diff
@@ -0,0 +1,172 @@
+        RebalanceDecision rebalanceDecision = allocator.rebalanceShard(unassignedShard, newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.emptyList()), clusterState));
+        assertSame(RebalanceDecision.NOT_TAKEN, rebalanceDecision);
+    }
+
+    public void testRebalanceNotAllowedDuringPendingAsyncFetch() {
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomBoolean(), ShardRoutingState.STARTED);
+        ShardRouting unassignedShard = clusterState.routingTable().index("idx").shard(0).primaryShard();
+        RoutingAllocation routingAllocation = newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.emptyList()), clusterState);
+        routingAllocation.setHasPendingAsyncFetch();
+        RebalanceDecision rebalanceDecision = allocator.rebalanceShard(unassignedShard, routingAllocation);
+        assertNotNull(rebalanceDecision.getCanRebalanceDecision());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertEquals("cannot rebalance due to in-flight shard store fetches", rebalanceDecision.getFinalExplanation());
+        assertNull(rebalanceDecision.getNodeDecisions());
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testRebalanceNotAllowed() {
+        AllocationDecider noRebalanceDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
+                return randomFrom(Decision.NO, Decision.THROTTLE);
+            }
+        };
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomBoolean(), ShardRoutingState.STARTED);
+        ShardRouting shard = clusterState.routingTable().index("idx").shard(0).primaryShard();
+        RebalanceDecision rebalanceDecision = allocator.rebalanceShard(shard, newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.singleton(noRebalanceDecider)), clusterState));
+        assertNotEquals(Type.YES, rebalanceDecision.getCanRebalanceDecision().type());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertEquals("rebalancing is not allowed", rebalanceDecision.getFinalExplanation());
+        assertNull(rebalanceDecision.getNodeDecisions());
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testBalanceShard() {
+        AllocationDecider canAllocateDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                return Decision.YES;
```

Comment:
The issue is that `yesAllocationDeciders` returns an instance of `AllocationDeciders`, and I needed to create my own to add a decider that overrides `canRebalance`.